### ISSUE TITLE
False Positive of `unnecessary_fallible_conversions` for pyo3

### ISSUE THREAD

**Issue by Xuanwo** [rust-lang/rust-clippy#12039]
### Summary

rust-clippy with 1.75 could emit false positive of `unnecessary_fallible_conversions` for pyo3:

```rust
warning: use of a fallible conversion when an infallible one could be used
   --> bindings/python/src/file.rs:180:27
    |
180 |     pub fn __enter__(slf: Py<Self>) -> Py<Self> {
    |                           ^^ help: use: `From::from`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_fallible_conversions
    = note: `#[warn(clippy::unnecessary_fallible_conversions)]` on by default

warning: `opendal-python` (lib test) generated 1 warning
warning: `opendal-python` (lib) generated 1 warning (1 duplicate)
```

It's bit confused to raising `unnecessary_fallible_conversions` on API like this:

```rust
pub fn __enter__(slf: Py<Self>) -> Py<Self> {
    slf
}
```

`Py` is a transparent struct expose by pyo3: https://docs.rs/pyo3/latest/pyo3/struct.Py.html

### Lint Name

unnecessary_fallible_conversions

### Reproducer

I tried this code:

```rust
pub fn __enter__(slf: Py<Self>) -> Py<Self> {
    slf
}
```

I saw this happen:

```
warning: use of a fallible conversion when an infallible one could be used
   --> bindings/python/src/file.rs:180:27
    |
180 |     pub fn __enter__(slf: Py<Self>) -> Py<Self> {
    |                           ^^ help: use: `From::from`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_fallible_conversions
    = note: `#[warn(clippy::unnecessary_fallible_conversions)]` on by default

warning: `opendal-python` (lib test) generated 1 warning
warning: `opendal-python` (lib) generated 1 warning (1 duplicate)
```

I expected to see this happen: no warning

### Version

```text
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

**Comment by djc** [comment]
I'm also seeing this. I guess this happens within the `#[pymethods]` expansion, and this lint isn't correcting for macro-generated code?

**Comment by y21** [comment]
That's odd, all method lints already have a check if they're from a macro: https://github.com/rust-lang/rust-clippy/blob/174a0d7be6f18bf6470a46d107e4b3f0b870c315/clippy_lints/src/methods/mod.rs#L4001-L4005

I guess `#[pymethods]` uses the span of the callsite, instead of definition site, which makes clippy think that this is something that the user wrote 🤔

**Comment by djc** [comment]
@davidhewitt is there something on the PyO3 site that could be improved here?

**Comment by davidhewitt** [comment]
We've already "fixed" this (with an `allow` in the generated output) on PyO3 main. I'd intended to release 0.21 with this before 1.75 landed but we got caught up in some larger work which has delayed the release and I forgot about the time-sensitive nature of this. Let me look into a patch release for 0.20 asap, as 0.21 is realistically still a couple weeks off at best.

**Comment by davidhewitt** [comment]
https://github.com/PyO3/pyo3/pull/3564

**Comment by djc** [comment]
@davidhewitt awesome as always, thanks!

**Comment by davidhewitt** [comment]
Patch release is out, @Xuanwo please let me know if there's still an issue after upgrading.

**Comment by Xuanwo** [comment]
> Patch release is out, @Xuanwo please let me know if there's still an issue after upgrading.

Great! The patch resolved my issue.
